rfc/project-cutover.md

```diff
@@ -0,0 +1,109 @@
+
+Non-Goals:
+* Change *how* the artifacts are built or published - just where they are done. including:
+  * Start using Github Package Repo
+  * Use Google Cloud Soure Repo during container builds
+  * Use of Jenkins to orchestrate the release process
+
+## Motivation and Rationale
+
+_What is the current state of the world?_
+* Docker containers are orchestrated from a Jenkins instance in `spinnaker-community` and actually built, stored, and served from the `spinnaker-marketplace` project.
+  * A nightly BOM and each release BOM is generated and stored in the `halconfig` bucket, along with copies of all microservice config files from their respective `halconfig/` directories in each repo.
+
+* Halyard, spin, and sponnet releases are all orchested from the same Jenkins instance and are stored in their respective locations in `spinnaker-artifacts`
+
+* `spinnakerbot` can be [built and deployed](https://github.com/spinnaker/spinnaker/blob/master/spinbot/Makefile#L11) by any developers with Google Container Builder (GCB) priviledges on the `spinnaker-marketplace` project.
+
+_Why is this change being proposed?_
+
+* Transparency in the process, remove Google as the sole owner/operator/bottleneck of the whole thing.
+
+_Who have you identified as early customers of this change?_
+
+* End-users and any new non-Google engineers that will be involved in the expanded nightly build-cop and release-manager duties.
+
+## Timeline
+
+_What is the proposed timeline for the implementation?_
+
+ASAP
+
+## Design
+
+_What exactly are you doing?_
+
+
+### `halconfig` bucket
+
+GCS buckets exist in a global namespace, and cannot be directly transferred between projects.
+
+Two possible approaches:
+1. (Preferred): Copy all data from bucket `halconfig` in `spinnaker-marketplace` to `halconfig-temp` in `spinnaker-community`. Then delete the old bucket and recreate it in the new project.
+2. Copy all data to `halconfig-v2` in the new project.
+  * Update all Halyard config to use this new bucket
+  * Update build and release config to start publishing here
+  * Update any documentation on spinnaker.io
+
+Approach 1 comes with some downtime as the data is transferred and buckets get renamed. Also comes with very slight risk of someone else grabbing the bucket name inbetween when the old bucket is deleted and the new one with the same name is created.
+
+`spinnaker-artifacts` would most likely follow the same approach
+
+### Microservice Docker registry
+
+Each BOM references the default Docker registry to find each container (e.g. `artifactSources.dockerRegistry: gcr.io/spinnaker-marketplace`). Using approach #1 for the `halconfig` bucket would mean BOMs would be served from the `spinnaker-community` project but Docker/Kubernetes clusters would pull from GCR in `spinnaker-marketplace`. Google would have to maintain this project for approximately 6 months after the cutover (8 weeks * last 3 supported release = 24 weeks or about 6 months).
```

Comment:
Are there any significant drawbacks to keeping the containers in the old project longer than 6 months (perhaps monitoring to see how often they are pulled)?

I ask because, while we support only 3 releases, I think that a lot of users do lag behind our officially-supported releases.  What will be the failure mode for them if we shut down this old project (and remove access to the containers)?  I think a worst-case scenario is that someone running an old version of Spinnaker makes some change to their cluster (ex: node pool upgrade) that requires an image re-pull that fails and leaves them with downtime and an unclear path to resolve.

For that reason, I'd probably advocate keeping images at their old location around quite a bit longer than the "officially supported" window.

Reply:
Totally agree - I have no intention to actively delete any of the old images until long after we've stabilized on the new project.